### PR TITLE
`Error` transport via `StatusRuntimeException`

### PR DESCRIPTION
**Pull request by dmytro-grankin**
Previously a client got an additional information about an occurred error from a `StatusRuntimeException` cause. It was fine until the [fix](https://github.com/grpc/grpc-java/issues/1716).

The following changes resolve the issue:
- Now a `StatusRuntimeException` additionally includes a metadata, containing an occurred `Error`.
- Introduced new client API to extract an occurred `Error` from a `StatusRuntimeException`.

New API usage:

```java
// Code in a StreamObserver.
@Override
public void onError(Throwable t) {
    final Optional<Error> error = Errors.fromStreamError(t);
}
```

### CODE REVIEW

**Comment by dmytro-grankin** [comment]
@armiol PTAL.

**Comment by armiol** [comment]
@dmytro-grankin 
```
final Optional<Error> error = Errors.fromStatus(t);
```
Why `fromStatus`? It's `Throwable`, isn't it?

**Comment by dmytro-grankin** [comment]
@armiol PTAL.

File: client/src/main/java/org/spine3/base/Errors.java [hunk 1]
```diff
@@ -32,14 +36,17 @@
  */
 public class Errors {
 
-    private Errors() {}
+    // Prevent instantiation of this utility class.
```
Comment (armiol) [inline client/src/main/java/org/spine3/base/Errors.java]: This should go inside the ctor.

File: client/src/main/java/org/spine3/base/Errors.java [hunk 2]
```diff
@@ -51,10 +58,39 @@ public static Error fromThrowable(Throwable throwable) {
...
+                                  .setType(throwable.getClass()
+                                                    .getName())
                                   .setMessage(message)
                                   .setStacktrace(Throwables.getStackTraceAsString(throwable))
                                   .build();
         return result;
     }
+
+    /**
+     * Extracts an {@link Error} either from the {@link StatusRuntimeException}
```
Comment (armiol) [inline client/src/main/java/org/spine3/base/Errors.java]:
```
Extracts a {@linkplain Error system error} from the {@linkplain io.grpc.stub.StreamObserver#onError(Throwable) Throwable}, received on a client-side as a result of a failed gRPC call to server-side routines.

<p>The {@code Error} is extracted from the trailer metadata of either {@link StatusRuntimeException} or {@link StatusException} only.

<p>If any other exception is passed, {@code Optional.absent()} is returned.

@param ...
@return ...
```

File: client/src/main/java/org/spine3/base/MetadataConverter.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import io.grpc.Metadata;
+import io.grpc.Metadata.Key;
+import org.spine3.annotations.Internal;
+import org.spine3.util.Exceptions;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.Metadata.BINARY_BYTE_MARSHALLER;
+
+/**
+ * Serves as converter from {@link Error} to {@link Metadata} and vice versa.
```
Comment (armiol) [inline client/src/main/java/org/spine3/base/MetadataConverter.java]: `as a converter`

File: client/src/main/java/org/spine3/base/MetadataConverter.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+    private static final String ERROR_KEY_NAME = "Spine-Error-bin";
+
+    /**
+     * The {@link Metadata.Key} to store and get an {@link Error} from a {@link Metadata}.
+     */
+    @VisibleForTesting
+    static final Key<byte[]> KEY = Key.of(ERROR_KEY_NAME, BINARY_BYTE_MARSHALLER);
+
+    // Prevent instantiation of this utility class.
```
Comment (armiol) [inline client/src/main/java/org/spine3/base/MetadataConverter.java]: Same as in similar case above.

File: client/src/test/java/org/spine3/base/ErrorsShould.java [hunk 5]
```diff
@@ -56,6 +61,43 @@ public void create_Error_by_Throwable() {
...
+
+        assertEquals(expectedError, Errors.fromStreamError(statusRuntimeException)
+                                          .get());
+    }
+
+    @Test
+    public void return_Error_extracted_form_StatusException_metadata() {
+        final Error expectedError = Error.getDefaultInstance();
+        final Metadata metadata = MetadataConverter.toMetadata(expectedError);
+        final StatusException statusException = Status.INVALID_ARGUMENT.asException(metadata);
```
Comment (armiol) [inline client/src/test/java/org/spine3/base/ErrorsShould.java]: Please have `Status.INVALID_ARGUMENT` statically imported.

File: client/src/test/java/org/spine3/base/ErrorsShould.java [hunk 6]
```diff
@@ -56,6 +61,43 @@ public void create_Error_by_Throwable() {
...
+        assertEquals(expectedError, Errors.fromStreamError(statusException)
+                                          .get());
+    }
+
+    @Test
+    public void return_absent_if_passed_Throwable_is_not_status_exception() {
+        final String msg = "Neither a StatusException nor a StatusRuntimeException.";
+        final Exception exception = new Exception(msg);
+
+        assertEquals(Optional.absent(), Errors.fromStreamError(exception));
```
Comment (armiol) [inline client/src/test/java/org/spine3/base/ErrorsShould.java]: `assertFalse(Errors.fromStreamError(exception).isPresent())`

File: client/src/main/java/org/spine3/base/Errors.java (hunk 1 above)
Comment (dmytro-grankin) [inline client/src/main/java/org/spine3/base/Errors.java]: Done.

File: client/src/main/java/org/spine3/base/MetadataConverter.java (hunk 4 above)
Comment (dmytro-grankin) [inline client/src/main/java/org/spine3/base/MetadataConverter.java]: Done.

File: client/src/main/java/org/spine3/base/Errors.java (hunk 2 above)
Comment (dmytro-grankin) [inline client/src/main/java/org/spine3/base/Errors.java]: Done.

File: client/src/main/java/org/spine3/base/MetadataConverter.java (hunk 3 above)
Comment (dmytro-grankin) [inline client/src/main/java/org/spine3/base/MetadataConverter.java]: Done.

File: client/src/test/java/org/spine3/base/ErrorsShould.java (hunk 5 above)
Comment (dmytro-grankin) [inline client/src/test/java/org/spine3/base/ErrorsShould.java]: Done.

File: client/src/test/java/org/spine3/base/ErrorsShould.java (hunk 6 above)
Comment (dmytro-grankin) [inline client/src/test/java/org/spine3/base/ErrorsShould.java]: Done.

**Comment by dmytro-grankin** [comment]
@armiol PTAL.

File: client/src/main/java/org/spine3/base/Errors.java [hunk 7]
```diff
@@ -51,10 +58,39 @@ public static Error fromThrowable(Throwable throwable) {
...
+
+    /**
+     * Extracts a {@linkplain Error system error} from the
+     * {@linkplain io.grpc.stub.StreamObserver#onError(Throwable) Throwable},
+     * received on a client-side as a result of a failed gRPC call to server-side routines.
+     *
+     * <p>The {@code Error} is extracted from the trailer metadata of
+     * either {@link StatusRuntimeException} or {@link StatusException} only.
+     *
+     * <p>If any other exception is passed, {@code Optional.absent()} is returned.
```
Comment (armiol) [inline client/src/main/java/org/spine3/base/Errors.java]: Let's make it even clearer:
```
<p>If any other type of {@code Throwable} is passed, ...
```

**Pull request merged**